src/Keboola/S3Extractor/S3ExceptionConverter.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+namespace Keboola\S3Extractor;
+
+use Aws\S3\Exception\S3Exception;
+use GuzzleHttp\Exception\ClientException;
+
+class S3ExceptionConverter
+{
+    /**
+     * @param S3Exception $e
+     * @throws Exception
+     * @throws S3Exception
+     */
+    public static function resolve(S3Exception $e): void
+    {
+        self::handleInvalidCredentials($e);
+        self::handleSlowDown($e);
+        self::handleBaseUserErrors($e);
+
+        throw $e;
+    }
+
+    /**
+     * @param S3Exception $e
+     * @throws Exception
+     */
+    private static function handleInvalidCredentials(S3Exception $e): void
+    {
+        if ($e->getStatusCode() === 403) {
+            throw new Exception('Invalid credentials or permissions.', $e->getCode(), $e);
+        }
+    }
+
+    /**
+     * @param S3Exception $e
+     * @throws Exception
+     */
+    private static function handleSlowDown(S3Exception $e): void
+    {
+        if ($e->getStatusCode() === 503) {
+            throw new Exception('Error 503 Slow Down: The number of requests to the S3 bucket is very high, please check your bucket limit.', $e->getCode(), $e);
+        }
+    }
+
+    /**
+     * @param S3Exception $e
+     * @throws Exception
+     */
+    private static function handleBaseUserErrors(S3Exception $e): void
+    {
+        if (false === in_array($e->getStatusCode(), [400, 401, 404], true)) {
+            return;
```

Comment:
toto mi dalo dost zabrat - proc to tu nevyhazuje vyjimku? - aha no vyhazuje, ale uplne nekde jinde

Reply:
Přišlo mi čitelnější než předchozí kód. Ale oka, co říkáš na switch https://github.com/keboola/aws-s3-extractor/pull/39/commits/70293639461933a34a0d88dd69d3670a032874db?

Reply:
Extrahovat exception handler jsem musel, jelikož bych to neměl jak testovat. Původně jsem kód jen přesunul, ale codeclimate hlásil cyclomatic complexity 10, s tím že je povoleno max 5. Bohužel to není v historii https://codeclimate.com/github/keboola/aws-s3-extractor/pull/39#issues. Nicméně tím se vysvětluje refactoring, který byl v rámci odstranění chyby podle mě potřeba.

Klidně to rozdělím, to není problém. Jen bych chtěl znát důvod, abychom byli pro příště synchronní. Souhlasíš, že v případě nahlášené chyby codeclimatem je validní udělat v PR při fixu i refactoring? Nebo ti jde striktně o historii gitu v rámci merge commits?

Reply:
> refactoring, který byl v rámci odstranění chyby podle mě potřeba.

refactoring = zmena kodu bez vnejsiho vlivu :) prijde mi, ze jestli ho potrebujes, tak to udelej napred - nic tomu tedy nebrani a urcite je to zmena k lepsimu (a ten switch jeste lip). Nicmene pokud to spojis, tak pak blbe zkouma co se kvuli cemu menilo a proc a jak (jak ted, tak historicky). Takze za me - nedelat soucasne (pokud je to mozny pochopitelne). Kdyztak -> #best_practices